### DISCUSSION TITLE
Timer with Line Trace

### DISCUSSION THREAD

**Discussion by germj** [nxrighthere/UnrealCLR#62] [Help] [answered]
Is there a way to create a Timer that would pulse every 0.1 seconds, and have a line trace attached to an actor with a debug draw.

I’ve read through the documentation, and don’t see these pieces exposed as of yet.

**Comment by nxrighthere** [comment, marked as answer]
A built-in timer implementation is not provided yet, but it's easy to implement a custom one using frame ticking. For line tracing see [this](https://github.com/nxrighthere/UnrealCLR/blob/master/API/World.md#methods) functionality and just pass the position of an actor to it. To debug line tracing use [`Debug.DrawLine()`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/Debug-DrawLine(Vector3_Vector3_Color_bool_float_byte_float).md). Also, see [this](https://github.com/nxrighthere/UnrealCLR/blob/master/Source/Managed/Tests/SpatialQueries.cs#L32) test.

**Reply by germj** [reply]
Hi, I got everything working except the frame tick. 

Would you be able to provide a sample. I’ll post the code that I used. Really appreciate the help.

**Reply by nxrighthere** [reply]
You can implement a timer in several ways:

1.  Using [`World.Time`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/World-Time.md) to get time difference between frames and invoke functions in any preferable approach (for example, you can use callback references stored in a collection with timestamps)
2. Set a time variable in milliseconds for a particular callback and subtract [`World.DeltaTime`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/World-DeltaTime.md) every frame from it, till the time reaches <= 0.0 milliseconds

Take a look at how timers usually implemented in Unity, you will find several helpful approaches.

**Reply by germj** [reply]
What would the impact be of using C# timers? 
https://stackoverflow.com/questions/12535722/what-is-the-best-way-to-implement-a-timer

**Reply by nxrighthere** [reply]
As far as I know, it's based on multi-threading while you have to dispatch callbacks on the main thread.

**Comment by germj** [comment]
Hi, I got it to work. Is it possible to have multiple classes running.

Instead of using a switch, I'd like these to run at the same time.
			runningSystem = testSystem  switch{
				TestSystems.Brain_ML => new AStar(),
				TestSystems.Grid_ML => new Grid(),
				_ => throw new Exception("Unknown system")
			};

___________________________________________________________
C# Timer Class running in CLR

        private static System.Timers.Timer aTimer;
        private static float timer;

        private static void SetTimer()
        {
            // Create a timer with a two second interval.
            aTimer = new System.Timers.Timer(1000);
            // Hook up the Elapsed event for the timer. 
            aTimer.Elapsed += OnTimedEvent;
            aTimer.AutoReset = true;
            aTimer.Enabled = true;
        }

Make sure you end the timer prior to closing out.

        public void OnEndPlay()
        {

            aTimer.Stop();
            aTimer.Dispose();
            Debug.ClearOnScreenMessages();
        }

**Reply by nxrighthere** [reply]
You are free to organize systems execution in any way you like, be it delegates, interfaces, or functions. Just use world events as the execution root.

Are you sure that your timer is not violating thread-safety and events dispatched on the main thread? Because if not, then it will cause state/memory corruption, crashes, and other bad things.
